app/models/user.rb

```diff
@@ -569,33 +569,11 @@ def send_welcome_notification
   end
 
   def set_username
-    set_temp_username if username.blank?
-    self.username = username&.downcase
+    self.username = username&.downcase&.presence || generate_username
   end
 
-  # @todo Should we do something to ensure that we don't create a username that violates our
```

Comment:
[This code has literally been here since the initial commit](https://github.com/forem/forem/commit/301c6080e3f64139ea339d37d8c151521bc07177#diff-9802ca3c9c4cf89904fd44bc114e35ebdf2c5dd3d5b645491e2b253e1afef29bR261-R286) and this is a great refactoring, thank you! 

One minor suggestion, as a way of having a bit of a separation of concerns between User and the new service, what if, since the Generator service specifically needs the auth-provider temporary usernames, instead of passing in user, what if we passed in those auth-provider-generated usernames? The Generator service could then be more focused on normalizing those usernames (verifying uniqueness, etc.) with fewer dependencies on `User` or `Authentication::Providers`.

Possibly something like:

```ruby
def auth_provider_usernames
  attributes.with_indifferent_access.slice(*Authentication::Providers.username_fields).values.compact
end
```